sdk/src/compressed_account.rs

```diff
@@ -1,421 +1,150 @@
+                        data: Some(data),
+                    },
+                    merkle_tree_index,
+                }))
             }
-            Self::Mut(light_mut_account) => {
-                let account =
-                    light_mut_account.output_compressed_account(program_id, remaining_accounts)?;
-                Ok(Some(account))
-            }
-            Self::Close(_) => Ok(None),
-        }
-    }
-}
-
-impl<T> Deref for LightAccount<T>
-where
-    T: AnchorDeserialize + AnchorSerialize + Clone + DataHasher + Discriminator,
-{
-    type Target = T;
-
-    fn deref(&self) -> &Self::Target {
-        match self {
-            Self::Init(light_init_account) => &light_init_account.output_account,
-            Self::Mut(light_mut_account) => &light_mut_account.output_account,
-            Self::Close(light_close_account) => &light_close_account.input_account,
-        }
-    }
-}
-
-impl<T> DerefMut for LightAccount<T>
-where
-    T: AnchorDeserialize + AnchorSerialize + Clone + DataHasher + Discriminator,
-{
-    fn deref_mut(&mut self) -> &mut Self::Target {
-        match self {
-            Self::Init(light_init_account) => &mut light_init_account.output_account,
-            Self::Mut(light_mut_account) => &mut light_mut_account.output_account,
-            Self::Close(light_close_account) => &mut light_close_account.input_account,
-        }
-    }
-}
-
-pub struct LightInitAccount<T>
-where
-    T: AnchorDeserialize + AnchorSerialize + Clone + DataHasher + Discriminator,
-{
-    output_account: T,
-    address_seed: Option<[u8; 32]>,
-    merkle_context: PackedMerkleContext,
-    address_merkle_context: PackedAddressMerkleContext,
-    address_merkle_tree_root_index: u16,
-}
-
-impl<T> LightInitAccount<T>
-where
-    T: AnchorDeserialize + AnchorSerialize + Clone + Default + DataHasher + Discriminator,
-{
-    pub fn new(
-        merkle_context: &PackedMerkleContext,
-        address_merkle_context: &PackedAddressMerkleContext,
-        address_merkle_tree_root_index: u16,
-    ) -> Self {
-        let output_account = T::default();
-
-        Self {
-            output_account,
-            address_seed: None,
-            merkle_context: *merkle_context,
-            address_merkle_context: *address_merkle_context,
-            address_merkle_tree_root_index,
-        }
-    }
-
-    pub fn set_address_seed(&mut self, address_seed: [u8; 32]) {
-        self.address_seed = Some(address_seed);
-    }
-
-    pub fn new_address_params(&self) -> NewAddressParamsPacked {
-        NewAddressParamsPacked {
-            seed: self.address_seed.unwrap(),
-            address_merkle_tree_account_index: self
-                .address_merkle_context
-                .address_merkle_tree_pubkey_index,
-            address_queue_account_index: self.address_merkle_context.address_queue_pubkey_index,
-            address_merkle_tree_root_index: self.address_merkle_tree_root_index,
+            None => Ok(None),
```

Comment:
This should throw an error, since we want the output Merkle tree to be set explicitly to avoid the trap of compressed accounts being stuck in the same tree.

Reply:
This branch is about handling `close` accounts (only input, no output=. How would you handle them differently when returning an error here?

I think that our helpers for creating `LightAccontMeta` are already simple enough that people shouldn't make that mistake, unless they use `LightAccountMeta::new_close`. If people construct meta by hand, then I think we should be fine with putting some more responsibility on them and assume they'd be able to debug the trap situation you're mentioning.

Another reason reason why I'm returning `None`, not `Err` in this case, is that I'm using `input_compressed_account` and `output_compressed_account` here in a loop:

https://github.com/Lightprotocol/light-protocol/blob/33d6683553ac43dadb01c7444a29a83242530cc9/sdk/src/verify.rs#L291-L303

And overall, my meaning behind having `Result<Option<OutputCompressedAccount>>` type is that:

- `Ok(Some)` means that the account is `init` or `mut` and we successfully generated the account.
- `Err` means that the account is `init` or `mut`, but there was an error with hashing or serialization.
- `Ok(None)` means that the account is `close`. The loop in `verify` wrappers will just not include such account in the outputs.

I think handling the last case without `None` would be very annoying.